### ISSUE TITLE
Add support for AWS Instance metadata version 2 (IMDSv2)

### ISSUE THREAD

**Issue by castrapel** [osquery/osquery#6165]
### What new feature do you want?

Add support for version 2 of the instance metadata service provided by AWS. ( https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html ).  

Instance metadata version 2 (IMDSv2) uses session-oriented requests. With session-oriented requests, you utilize a specific set of commands and a specified duration of up to six hours to complete the request. During the specified duration, you can use the same session identifier for all of your requests. After the specified duration expires, you must create a new session.

### How is this new feature useful?

Companies may want to restrict usage of IMDSv1 while still using osquery to query the metadata service.

IMDSv1 can be restricted with the following command:

`aws ec2 modify-instance-metadata-options --instance-id <instance_id> --http-tokens required --http-endpoint enabled`

### How can this be implemented?

I presume by modifying HTTP requests utilizing the kEc2MetadataUrl variable:
https://github.com/osquery/osquery/search?q=kEc2MetadataUrl&unscoped_q=kEc2MetadataUrl , adding requests to get a token, and making requests to kEc2MetadataUrl with that token. You can choose to reuse a token or create a new token with every request. In a small number of requests, it might be easier to generate and use a token for one-time use. But for efficiency, you can specify a longer duration for the token and reuse it rather than having to write a PUT request every time you need to request instance metadata. 

The following example of a Linux shell script uses curl and IMDSv2 to retrieve instance metadata on an EC2 instance. The example command creates a session of six hours (21,600 seconds), stores the session token header in a variable named token, and then requests the top-level metadata items using the token.

[ec2-user ~]$ TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"` \
&& curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data/

After you've created a token, you can reuse it until it expires. In the following example command, the token is reused:

[ec2-user ~]$ curl -H "X-aws-ec2-metadata-token: $TOKEN" -v http://169.254.169.254/latest/meta-data/

**Comment by nyanshak** [comment]
https://github.com/osquery/osquery/commit/e2b3598bacf5e816d931f8063504399c780afac2#diff-0a259f9ce19b48500001188080d56dd372abeca74fe87340723a07ec718f79c3 - In this commit, the AWS SDK was updated to 1.7.230. This SDK version should add proper support for EC2 metadata v2.

This update went out with 4.6.0.

```
# from tcpdump on osquery 4.6.0

	GET /latest/meta-data/iam/security-credentials/test-instance HTTP/1.1
	host: 169.254.169.254
	user-agent: aws-sdk-cpp/1.7.230 Linux/<kernel-version>-aws x86_64 Clang/9.0.1
	x-aws-ec2-metadata-token:
```

It looks to me like this has been solved in 4.6.0 and this issue can be closed. Can you confirm?

**Comment by nyanshak** [comment]
OK - did some code review with @suppandi and @Gabriel Munoz in osquery slack.

The functionality that uses the AWS SDK works fine. The behavior is - if v1 is still enabled, continue using v1 metadata. If the instance is v2 only, it will use v2 metadata endpoint. So for example - assuming roles / sending to Kinesis will work. However there is other functionality within osquery that does not seem to work correctly with v2.

With metadata v2 _required_, osquery 4.7.0, `ec2_instance_tags`:
```
Using a virtual database. Need help, type '.help'
osquery> select * from ec2_instance_tags;
osquery> select * from ec2_instance_tags;
osquery> select * from ec2_instance_tags;
```

With metadata v2 _optional_, osquery 4.7.0, `ec2_instance_tags`:
osquery> select * from ec2_instance_tags;
# sometimes I'm getting this:
I0421 14:52:12.583938 17383 aws_util.cpp:397] EC2 instance ID: i-1234. Region: ap-southeast-2
I0421 14:52:12.616344 17383 ec2_instance_tags.cpp:50] Error getting EC2 instance tags: The request must contain the parameter AWSAccessKeyId
osquery> select * from ec2_instance_tags;
# other times when I call the table, I get this error output 5 times, followed by results:
E0421 14:52:17.444041 17383 aws_util.cpp:202] Exception making HTTP request to URL (http://169.254.169.254/latest/api/token): field value too large
+---------------------+----------------+---------------+
| instance_id         | key            | value         |
+---------------------+----------------+---------------+
| i-1234                          | some_key  | some_value |
```

With metadata v2 _optional_, osquery 4.7.0, `ec2_instance_metadata, running SELECT against ec2_instance_metadata works fine:

```
osquery> select instance_type from ec2_instance_metadata;
+---------------+
| instance_type |
+---------------+
| t2.micro      |
+---------------+
```

With metadata v2 _required_, osquery 4.7.0, `ec2_instance_metadata` table does not work:

```
Using a virtual database. Need help, type '.help'
osquery> select instance_type from ec2_instance_metadata;
osquery> select instance_type from ec2_instance_metadata;
osquery>

# If I start osquery with *optional* v2, run query, then disable:
#   this query is run with *optional v2*
osquery> select instance_type from ec2_instance_metadata;
+---------------+
| instance_type |
+---------------+
| t2.micro      |
+---------------+
# same osqueryi process, after *requiring* v2:
osquery> select instance_type from ec2_instance_metadata;
I0421 15:06:29.069038 17478 ec2_instance_metadata.cpp:146] Unexpected HTTP response for: dynamic/instance-identity/document Status: 401
I0421 15:06:29.069214 17478 ec2_instance_metadata.cpp:202] Could not parse JSON from dynamic/instance-identity/document: <unspecified file>(1): expected value
I0421 15:06:29.070478 17478 ec2_instance_metadata.cpp:146] Unexpected HTTP response for: meta-data/iam/info Status: 401
I0421 15:06:29.070626 17478 ec2_instance_metadata.cpp:202] Could not parse JSON from meta-data/iam/info: <unspecified file>(1): expected value
I0421 15:06:29.072149 17478 ec2_instance_metadata.cpp:146] Unexpected HTTP response for: meta-data/mac Status: 401
I0421 15:06:29.073230 17478 ec2_instance_metadata.cpp:146] Unexpected HTTP response for: meta-data/local-hostname Status: 401
I0421 15:06:29.074199 17478 ec2_instance_metadata.cpp:146] Unexpected HTTP response for: meta-data/public-keys/0/openssh-key Status: 401
I0421 15:06:29.075866 17478 ec2_instance_metadata.cpp:146] Unexpected HTTP response for: meta-data/reservation-id Status: 401
I0421 15:06:29.076941 17478 ec2_instance_metadata.cpp:146] Unexpected HTTP response for: meta-data/security-groups Status: 401
+---------------+
| instance_type |
+---------------+
|               |
+---------------+
```

---

tl;dr:
* `ec2_instance_metadata` and `ec2_instance_tags` do not seem to work with metadata v2 required
* `ec2_instance_tags` table seems a bit flaky / logging errors even when metadata v1 is available
* Behavior when switching from v1 to v2 with a long-running osquery process that queries those tables seems to be different than a newly started osquery process that has never queried those tables.

**Comment by directionless** [comment]
> The behavior is - if v1 is still enabled, continue using v1 metadata. If the instance is v2 only, it will use v2 metadata endpoint

I wonder if it would make sense to have a hidden column allowing the caller to request a version.
